File: src/commands/cloud_agent/codex.py
"""Detached Codex App Server bootstrap; requests and credentials travel over SSH."""
import base64
import fcntl
import hashlib
import http.client
import json
import os
from pathlib import Path
import re
import signal
import socket
import subprocess
import sys
import time

CODE_PORT = 4096
LEGACY_PORT = 8080


class SetupError(Exception):
    pass


def save(path, value):
    temporary = path.with_suffix(".tmp")
    with temporary.open("w") as stream:
        os.chmod(temporary, 0o600)
        json.dump(value, stream)
    temporary.replace(path)


def process_start(pid):
    try:
        fields = Path(f"/proc/{pid}/stat").read_text().rsplit(")", 1)[1].split()
        return None if fields[0] == "Z" else fields[19]
    except (OSError, IndexError):
        return None


def restored_state(state):
    current = os.environ.get("RAILWAY_FACTORY_VM_ID")
    return bool(current and state.get("vm_id") and state["vm_id"] != current)


def owned_process(state):
    if restored_state(state):
        return False
    pid, start = state.get("pid"), state.get("start")
    return isinstance(pid, int) and pid > 1 and start is not None and process_start(pid) == start


def probe(port, token=None, websocket=False):
    # Direct loopback probes bypass HTTP_PROXY. Health endpoints are intentionally
    # public; only a WebSocket upgrade tests the server's authentication boundary.
    connection = http.client.HTTPConnection("127.0.0.1", port, timeout=1)
    headers = {}
    if websocket:
        key = base64.b64encode(os.urandom(16)).decode()
        headers = {"Connection": "Upgrade", "Upgrade": "websocket",
                   "Sec-WebSocket-Key": key, "Sec-WebSocket-Version": "13"}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    try:
        connection.request("GET", "/" if websocket else "/readyz", headers=headers)
        response = connection.getresponse()
        if websocket and response.status == 101:
            expected = base64.b64encode(hashlib.sha1(
                (key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11").encode()
            ).digest()).decode()
            if response.getheader("Sec-WebSocket-Accept") != expected:
                return None
        return response.status
    except (OSError, http.client.HTTPException):
        return None
    finally:
        connection.close()


def healthy(port, token):
    return (probe(port) == 200 and probe(port, token, True) == 101
            and probe(port, websocket=True) == 401)


def port_available(port):
    try:
        # Some platforms let a reusable wildcard bind coexist with a listener
        # on a specific interface. Check loopback explicitly as well.
        for address in ("0.0.0.0", "127.0.0.1"):
            with socket.socket() as sock:
                sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
                sock.bind((address, port))
                sock.listen(1)
        return True
    except OSError:
        return False


def runtime(environment):
    try:
        help_result = subprocess.run(["codex", "app-server", "--help"], env=environment,
                                     capture_output=True, text=True, timeout=15, check=True)
        if not all(flag in help_result.stdout for flag in ("--ws-auth", "--ws-token-file", "--listen")):
            raise SetupError("Update Codex on this agent: authenticated App Server WebSockets are required.")
        version = subprocess.run(["codex", "--version"], env=environment,
                                 capture_output=True, text=True, timeout=15, check=True).stdout.strip()
    except (OSError, subprocess.SubprocessError):
        raise SetupError("Could not run Codex on this agent. Install a current @openai/codex release and retry.") from None
    match = re.fullmatch(r"codex-cli ([0-9]+\.[0-9]+\.[0-9]+(?:-[a-zA-Z0-9.-]+)?)", version)
    if not match:
        raise SetupError("Could not identify the remote Codex version.")
    return match[1]


def server_port(state):
    configured = os.environ.get("RAILWAY_CODE_PORT")
    if configured:
        if not configured.isascii() or not configured.isdecimal():
            raise SetupError("RAILWAY_CODE_PORT must be a valid code endpoint port.")
        configured = int(configured)
        if not 1024 <= configured <= 65535 or configured in (LEGACY_PORT, 8790):
            raise SetupError("RAILWAY_CODE_PORT must be 1024-65535, excluding the app and gateway ports.")
    # A live legacy server retains its endpoint. On a restored disk there is
    # no owned process: adopt the new VM's explicit code port, even when the
    # checkpoint contains pre-port launcher state. The API owns this setting.
    if owned_process(state):
        port = state.get("port", LEGACY_PORT)
    elif configured:
        port = configured
    elif state and not restored_state(state):
        port = state.get("port", LEGACY_PORT)
    else:
        port = CODE_PORT if os.environ.get(f"RAILWAY_PUBLIC_DOMAIN_{CODE_PORT}") else LEGACY_PORT
    if type(port) is not int or not 1024 <= port <= 65535 or port == 8790:
        raise SetupError("The saved Codex server has an unsupported port.")
    return port


def setup(request, home):
    os.umask(0o077)
    action = request.get("action", "start")
    if action not in ("start", "connect", "inspect"):
        raise SetupError("Unsupported Codex server action.")
    root = Path(home) / ".railway" / "codex"
    state_path = root / "server.json"
    if action == "inspect":
        state = json.loads(state_path.read_text()) if state_path.exists() else {}
        port = server_port(state)
        if (owned_process(state) and state.get("token") and state.get("directory")
                and healthy(port, state["token"])):
            return {"directory": state["directory"], "version": state.get("version")}
        return None
    if action == "connect" and not state_path.exists():
        raise SetupError("This agent has no managed Codex server. Run railway code --codex --agent <name> first.")
    root.mkdir(parents=True, exist_ok=True, mode=0o700)
    os.chmod(root, 0o700)
    with (root / "setup.lock").open("w") as lock:
        try:
            fcntl.flock(lock, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BlockingIOError:
            raise SetupError("Another Codex setup is running on this agent. Retry when it finishes.") from None
        state = json.loads(state_path.read_text()) if state_path.exists() else {}
        port = server_port(state)
        if action == "connect":
            if not state.get("directory") or not state.get("token"):
                raise SetupError("The saved Codex connection is incomplete. Rerun railway code --codex --agent <name>.")
            request = dict(request, directory=state["directory"], token=state["token"])
        directory = str(Path(request["directory"]).resolve(strict=True))
        if not Path(directory).is_dir():
            raise SetupError("Codex's working directory must be a directory.")
        domain = os.environ.get(f"RAILWAY_PUBLIC_DOMAIN_{port}")
        if not domain and port == LEGACY_PORT:
            domain = os.environ.get("RAILWAY_PUBLIC_DOMAIN")
        if not domain or not re.fullmatch(r"[a-zA-Z0-9.-]+", domain):
            raise SetupError(f"This agent has no valid public address for port {port}.")
        token = state.get("token") or request.get("token")
        if not isinstance(token, str) or not re.fullmatch(r"[a-zA-Z0-9_-]{43,}", token):
            raise SetupError("Codex requires a high-entropy URL-safe connection token.")

        reused = owned_process(state)
        if reused:
            if not healthy(port, token):
                raise SetupError(f"The managed Codex server is unhealthy. Check {root / 'server.log'} on the agent.")
            if state.get("directory") != directory:
                raise SetupError(f"Codex is already serving {state.get('directory')}. Use connect or --new for a different directory.")
        else:
            if not port_available(port):
                raise SetupError(f"Port {port} is occupied by another process. Use --new for a fresh agent.")
            environment = dict(os.environ, HOME=str(home))
            environment["PATH"] = f"{home}/.local/bin:" + environment.get("PATH", "/usr/local/bin:/usr/bin:/bin")
            version = runtime(environment)
            token_path = root / "server-token"
            token_path.write_text(token)
            os.chmod(token_path, 0o600)
            state = dict(token=token, directory=directory, version=version, port=port,
                         vm_id=os.environ.get("RAILWAY_FACTORY_VM_ID"))
            save(state_path, state)
            with (root / "server.log").open("w") as log:
                child = subprocess.Popen(
                    ["codex", "app-server", "--listen", f"ws://0.0.0.0:{port}",
                     "--ws-auth", "capability-token", "--ws-token-file", str(token_path)],
                    cwd=directory, env=environment, stdin=subprocess.DEVNULL,
                    stdout=log, stderr=subprocess.STDOUT, close_fds=True, start_new_session=True,
                )
            state.update(pid=child.pid, start=process_start(child.pid))
            save(state_path, state)
            deadline = time.monotonic() + 45
            try:
                while time.monotonic() < deadline:
                    if child.poll() is not None:
                        raise SetupError(f"Codex exited during startup. Check {root / 'server.log'} on the agent.")
                    if probe(port) == 200:
                        if not healthy(port, token):
                            raise SetupError("Codex did not enforce WebSocket authentication; stopped it.")
                        break
                    time.sleep(0.25)
                else:
                    raise SetupError(f"Codex startup timed out. Check {root / 'server.log'} on the agent.")
            except BaseException:
                if child.poll() is None:
                    os.killpg(child.pid, signal.SIGTERM)
                raise
        if reused and not state.get("vm_id") and os.environ.get("RAILWAY_FACTORY_VM_ID"):
            state["vm_id"] = os.environ["RAILWAY_FACTORY_VM_ID"]
            save(state_path, state)
        # Codex's native --remote parser requires an explicit port, including
        # the default TLS port. Keep it in the wire value (URL serializers omit it).
        return dict(url=f"wss://{domain}:443", token=token, directory=directory,
                    version=state["version"], reused=reused)


if __name__ == "__main__":
    try:
        result = setup(json.load(sys.stdin), Path.home())
        print("RAILWAY_CODEX_CONNECTION=" + json.dumps(result), flush=True)
    except (SetupError, OSError, ValueError, KeyError) as error:
        print(f"Codex setup failed: {error}", file=sys.stderr)
        sys.exit(1)

File: src/commands/cloud_agent/codex.rs
//! Authenticated Codex App Server on the cloud agent's public WebSocket endpoint.
use std::{process::Stdio, time::Duration};

use anyhow::{Context, Result, bail};
use futures_util::{SinkExt, StreamExt};
use reqwest_websocket::{Message, RequestBuilderExt};
use serde::Deserialize;
use serde_json::json;
use tokio::io::AsyncWriteExt;

use crate::commands::{
    code::{ConnectInfo, Prepared},
    ssh::native,
};
use crate::util::shell::shell_join;

pub(crate) mod bridge;
pub(crate) mod local;

const BOOTSTRAP: &str = include_str!("codex.py");
const RESULT_PREFIX: &str = "RAILWAY_CODEX_CONNECTION=";
pub(crate) const TOKEN_ENV: &str = "RAILWAY_CODEX_SERVER_TOKEN";

// No Debug: the connection contains a bearer token.
#[derive(Clone, Deserialize, serde::Serialize)]
pub(crate) struct Connection {
    pub url: String,
    pub token: String,
    pub directory: String,
    pub version: String,
    pub reused: bool,
}

async fn bootstrap(info: &ConnectInfo, request: serde_json::Value) -> Result<String> {
    let mut options = info.relay_opts.clone();
    if let Some(identity) = &info.identity {
        options.extend(["-i".into(), identity.to_string_lossy().into_owned()]);
    }
    options.extend(native::relay_port_args());
    let python = shell_join(&["python3".into(), "-c".into(), BOOTSTRAP.into()]);
    let mut child = tokio::process::Command::new("ssh")
        .args(options)
        .args(["-T", "-o", "BatchMode=yes", "-o", "ConnectTimeout=20", "--"])
        .arg(native::relay_destination(&info.ssh_target))
        .arg(shell_join(&["bash".into(), "-lc".into(), python]))
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true)
        .spawn()
        .context("Failed to start Codex setup over SSH")?;
    let mut stdin = child.stdin.take().context("SSH stdin unavailable")?;
    stdin.write_all(&serde_json::to_vec(&request)?).await?;
    drop(stdin);
    let timeout = if request["action"] == "inspect" {
        15
    } else {
        360
    };
    let output = tokio::time::timeout(Duration::from_secs(timeout), child.wait_with_output())
        .await
        .context("Codex setup timed out; rerun connect to check the server")??;
    if !output.status.success() {
        // stdout may contain credentials even if SSH subsequently failed.
        bail!(
            "Codex setup failed ({}): {}",
            output.status,
            String::from_utf8_lossy(&output.stderr).trim()
        );
    }
    String::from_utf8_lossy(&output.stdout)
        .lines()
        .find_map(|line| line.strip_prefix(RESULT_PREFIX))
        .map(str::to_owned)
        .context("SSH returned no Codex setup result")
}

pub(crate) async fn start_prepared(
    prepared: &Prepared,
    directory: &str,
    token: &str,
) -> Result<Connection> {
    let info = ConnectInfo {
        ssh_target: prepared.ssh_target.clone(),
        identity: prepared.identity.clone(),
        relay_opts: prepared.relay_opts.clone(),
    };
    connect_with(&info, json!({"directory": directory, "token": token})).await
}

pub(crate) async fn inspect(info: &ConnectInfo) -> Result<bool> {
    let result = bootstrap(info, json!({"action": "inspect"})).await?;
    let value: serde_json::Value = serde_json::from_str(&result)?;
    Ok(value["directory"]
        .as_str()
        .is_some_and(|dir| !dir.is_empty()))
}

pub(crate) async fn reconnect(info: &ConnectInfo) -> Result<Connection> {
    connect_with(info, json!({"action": "connect"})).await
}

async fn connect_with(info: &ConnectInfo, request: serde_json::Value) -> Result<Connection> {
    let connection: Connection = serde_json::from_str(&bootstrap(info, request).await?)
        .context("Invalid Codex connection result")?;
    validate_url(&connection.url)?;
    local::validate_version(&connection.version)?;
    verify_connection(&connection).await?;
    trust_directory(&connection, &connection.directory).await?;
    Ok(connection)
}

/// A short-lived authenticated App Server connection. All requests, including
/// notifications that precede their reply, have a bounded lifetime.
pub(super) struct Rpc {
    socket: reqwest_websocket::WebSocket,
    next_id: u64,
}

impl Rpc {
    pub(super) async fn connect(connection: &Connection) -> Result<Self> {
        let mut endpoint = validate_url(&connection.url)?;
        endpoint.set_scheme("https").expect("WSS can use HTTPS");
        Self::open(endpoint.as_str(), &connection.token).await
    }

    async fn open(endpoint: &str, token: &str) -> Result<Self> {
        tokio::time::timeout(Duration::from_secs(10), async {
            let client = reqwest::Client::builder()
                .redirect(reqwest::redirect::Policy::none())
                .build()?;
            let response = client
                .get(endpoint)
                .bearer_auth(token)
                .upgrade()
                .send()
                .await?;
            response.error_for_status_ref()?;
            let mut rpc = Self {
                socket: response.into_websocket().await?,
                next_id: 0,
            };
            rpc.call(
                "initialize",
                json!({
                    "clientInfo": {"name": "railway_cli", "version": env!("CARGO_PKG_VERSION")},
                    "capabilities": {"experimentalApi": true}
                }),
            )
            .await?;
            rpc.socket
                .send(Message::Text(json!({"method": "initialized"}).to_string()))
                .await?;
            Ok(rpc)
        })
        .await
        .context("Codex connection timed out")?
    }

    pub(super) async fn call(
        &mut self,
        method: &str,
        params: serde_json::Value,
    ) -> Result<serde_json::Value> {
        self.next_id += 1;
        let id = self.next_id;
        tokio::time::timeout(Duration::from_secs(10), async {
            self.socket
                .send(Message::Text(
                    json!({"id": id, "method": method, "params": params}).to_string(),
                ))
                .await?;
            while let Some(frame) = self.socket.next().await {
                match frame? {
                    Message::Text(text) => {
                        let value: serde_json::Value = serde_json::from_str(&text)?;
                        if value["id"] != id {
                            continue;
                        }
                        if let Some(error) = value.get("error") {
                            bail!(
                                "Codex {method} failed: {}",
                                error["message"].as_str().unwrap_or("App Server error")
                            );
                        }
                        return value
                            .get("result")
                            .cloned()
                            .context("Codex returned no result");
                    }
                    Message::Ping(data) => self.socket.send(Message::Pong(data)).await?,
                    _ => {}
                }
            }
            bail!("Codex disconnected during {method}")
        })
        .await
        .with_context(|| format!("Codex {method} timed out"))?
    }
}

/// Use the same remote config API as Codex's trust screen. The trust target may
/// be a repository root above cwd; trusting only cwd would leave that prompt up.
pub(crate) async fn trust_directory(connection: &Connection, directory: &str) -> Result<()> {
    let mut rpc = Rpc::connect(connection).await?;
    trust_with(&mut rpc, directory).await
}

async fn trust_with(rpc: &mut Rpc, directory: &str) -> Result<()> {
    let config = rpc
        .call(
            "config/read",
            json!({"cwd": directory, "includeLayers": true}),
        )
        .await?;
    let target = trust_target(&config, directory);
    if config["config"]["projects"][&target]["trust_level"] != "trusted" {
        rpc.call(
            "config/value/write",
            json!({
                "keyPath": format!("projects.{}.trust_level", serde_json::to_string(&target)?),
                "value": "trusted", "mergeStrategy": "replace"
            }),
        )
        .await?;
    }
    Ok(())
}

fn trust_target(config: &serde_json::Value, directory: &str) -> String {
    config["layers"]
        .as_array()
        .into_iter()
        .flatten()
        .rev()
        .filter(|layer| layer["name"]["type"] == "project")
        .find_map(|layer| {
            let reason = layer["disabledReason"].as_str()?;
            reason
                .split_once(", add ")
                .and_then(|(_, rest)| rest.rsplit_once(" as a trusted project in "))
                .map(|(path, _)| path)
                .or_else(|| {
                    layer["name"]["dotCodexFolder"]
                        .as_str()?
                        .strip_suffix("/.codex")
                })
        })
        .unwrap_or(directory)
        .to_string()
}

fn validate_url(value: &str) -> Result<url::Url> {
    let url = url::Url::parse(value).context("Invalid Codex public URL")?;
    if url.scheme() != "wss"
        || url.host_str().is_none()
        || !url.username().is_empty()
        || url.password().is_some()
        || url.path() != "/"
        || url.query().is_some()
        || url.fragment().is_some()
    {
        bail!("Codex's public address must be a WSS origin without embedded credentials");
    }
    Ok(url)
}

async fn verify_connection(connection: &Connection) -> Result<()> {
    let mut endpoint = validate_url(&connection.url)?;
    endpoint
        .set_scheme("https")
        .expect("WSS origin can use HTTPS");
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .timeout(Duration::from_secs(5))
        .build()?;
    let deadline = tokio::time::Instant::now() + Duration::from_secs(60);
    loop {
        if let Ok(response) = client.get(endpoint.join("readyz")?).send().await {
            if response.status().is_redirection() {
                bail!("Codex's public endpoint redirected unexpectedly");
            }
            if response.status().is_success() {
                break;
            }
        }
        if tokio::time::Instant::now() >= deadline {
            bail!("Codex started, but its public readiness check failed. Rerun connect to retry.");
        }
        tokio::time::sleep(Duration::from_secs(1)).await;
    }
    verify_websocket(&client, endpoint.as_str(), &connection.token).await
}

async fn verify_websocket(client: &reqwest::Client, endpoint: &str, token: &str) -> Result<()> {
    let unauthenticated = client.get(endpoint).upgrade().send().await?;
    if unauthenticated.status() != reqwest::StatusCode::UNAUTHORIZED {
        bail!("Codex's public endpoint did not reject an unauthenticated WebSocket connection");
    }
    let response = client
        .get(endpoint)
        .bearer_auth(token)
        .upgrade()
        .send()
        .await?;
    response.error_for_status_ref()?;
    let mut socket = response
        .into_websocket()
        .await
        .context("Opening authenticated Codex WebSocket")?;
    socket
        .send(Message::Text(
            json!({
                "id": 1, "method": "initialize", "params": {
                    "clientInfo": {"name": "railway_cli", "version": env!("CARGO_PKG_VERSION")}
                }
            })
            .to_string(),
        ))
        .await?;
    tokio::time::timeout(Duration::from_secs(10), async {
        while let Some(frame) = socket.next().await {
            if let Message::Text(text) = frame? {
                let value: serde_json::Value = serde_json::from_str(&text)?;
                if value["id"] != 1 {
                    continue;
                }
                if value.get("error").is_some() || !value["result"].is_object() {
                    bail!("Codex rejected the App Server initialization handshake");
                }
                socket
                    .send(Message::Text(json!({"method": "initialized"}).to_string()))
                    .await?;
                SinkExt::close(&mut socket).await?;
                return Ok(());
            }
        }
        bail!("Codex closed the connection before initialization")
    })
    .await
    .context("Codex initialization timed out")?
}

pub(crate) fn attach_args(connection: &Connection) -> Vec<String> {
    vec![
        // Railway pins the local client to its server. Codex's global-update
        // prompt can steal startup input and would break that version match.
        "-c".into(),
        "check_for_update_on_startup=false".into(),
        "--remote".into(),
        connection.url.clone(),
        "--remote-auth-token-env".into(),
        TOKEN_ENV.into(),
        "--cd".into(),
        connection.directory.clone(),
        "--ask-for-approval".into(),
        "never".into(),
        "--sandbox".into(),
        "danger-full-access".into(),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn trust_targets_the_disabled_repository_layer_including_quoted_paths() {
        let root = "/app/a \"project\"";
        let config = json!({"layers": [{"name": {"type": "project", "dotCodexFolder": format!("{root}/.codex")},
            "disabledReason": format!("Project config is disabled, add {root} as a trusted project in /home/me/.codex/config.toml")}]});
        assert_eq!(trust_target(&config, "/app/a \"project\"/child"), root);
        let key = format!(
            "projects.{}.trust_level = \"trusted\"",
            serde_json::to_string(root).unwrap()
        );
        let parsed: toml::Value = toml::from_str(&key).unwrap();
        assert_eq!(
            parsed["projects"][root]["trust_level"].as_str(),
            Some("trusted")
        );
        assert_eq!(trust_target(&json!({}), "/app"), "/app");
    }

    #[tokio::test]
    #[ignore = "requires a locally installed Codex App Server"]
    async fn real_codex_trust_persists_without_replacing_other_config() {
        let root = tempfile::tempdir().unwrap();
        let home = root.path().join("codex-home");
        let project = root.path().join("a project");
        std::fs::create_dir_all(&home).unwrap();
        std::fs::create_dir_all(project.join(".codex")).unwrap();
        std::fs::write(home.join("config.toml"), "# retained\nmodel = \"gpt-5\"\n").unwrap();
        std::fs::write(
            project.join(".codex/config.toml"),
            "model_reasoning_effort = \"high\"\n",
        )
        .unwrap();
        let listener = std::net::TcpListener::bind("127.0.0.1:0").unwrap();
        let address = listener.local_addr().unwrap();
        drop(listener);
        let mut server = tokio::process::Command::new("codex")
            .args(["app-server", "--listen", &format!("ws://{address}")])
            .env("CODEX_HOME", &home)
            .current_dir(&project)
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .kill_on_drop(true)
            .spawn()
            .unwrap();
        let deadline = tokio::time::Instant::now() + Duration::from_secs(15);
        let mut rpc = loop {
            match Rpc::open(&format!("http://{address}"), "").await {
                Ok(rpc) => break rpc,
                Err(error) => {
                    assert!(tokio::time::Instant::now() < deadline, "{error:#}");
                    tokio::time::sleep(Duration::from_millis(100)).await;
                }
            }
        };
        let directory = project
            .canonicalize()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        trust_with(&mut rpc, &directory).await.unwrap();
        trust_with(&mut rpc, &directory).await.unwrap();
        let stored = std::fs::read_to_string(home.join("config.toml")).unwrap();
        assert!(stored.contains("# retained"));
        let parsed: toml::Value = toml::from_str(&stored).unwrap();
        assert_eq!(parsed["model"].as_str(), Some("gpt-5"));
        assert_eq!(
            parsed["projects"][&directory]["trust_level"].as_str(),
            Some("trusted")
        );
        let effective = rpc
            .call("config/read", json!({"cwd":directory,"includeLayers":true}))
            .await
            .unwrap();
        rpc.call("thread/list", json!({"limit":100,"sortKey":"updated_at","archived":false,"sourceKinds":["cli","vscode","appServer","exec"]})).await.unwrap();
        assert!(
            effective["layers"]
                .as_array()
                .unwrap()
                .iter()
                .filter(|l| l["name"]["type"] == "project")
                .all(|l| l["disabledReason"].is_null())
        );
        server.kill().await.unwrap();
    }

    #[test]
    fn public_endpoint_requires_tls_and_separate_auth() {
        for invalid in [
            "ws://host",
            "https://host",
            "wss://user:secret@host",
            "wss://host/path",
            "wss://host?token=secret",
            "wss://host#fragment",
        ] {
            assert!(validate_url(invalid).is_err(), "{invalid}");
        }
        assert!(validate_url("wss://agent.example.com").is_ok());
    }

    #[cfg(unix)]
    #[test]
    fn bootstrap_preserves_server_identity_and_authentication() {
        let output = std::process::Command::new("python3")
            .arg(concat!(
                env!("CARGO_MANIFEST_DIR"),
                "/tests/codex_server.py"
            ))
            .output()
            .expect("python3 is required for Codex bootstrap tests");
        assert!(
            output.status.success(),
            "{}",
            String::from_utf8_lossy(&output.stderr)
        );
    }
}
